common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java

```diff
@@ -0,0 +1,814 @@
+                    }
+                }
+
+                lastBlockPos = blockPos;
+
+                if (!pauseDraw) {
+                    Vec3 rawLocation = point.vec3();
+                    int pathColor = point.color();
+                    consumer.vertex(lastMatrix, (float) rawLocation.x, (float) rawLocation.y, (float) rawLocation.z)
+                            .color(pathColor)
+                            .normal(0, 0, 1)
+                            .endVertex();
+                } else if (!sourceBatchEnded) {
+                    source.endBatch();
+                    sourceBatchEnded = true;
+                }
+            }
+            if (!sourceBatchEnded) {
+                for (ColoredPoint location : toRender) {
+                    Vec3 rawLocation = location.vec3();
+                    int pathColor = location.color();
+                    consumer.vertex(lastMatrix, (float) rawLocation.x, (float) rawLocation.y, (float) rawLocation.z)
+                            .color(pathColor)
+                            .normal(0, 0, 1)
+                            .endVertex();
+                }
+                source.endBatch();
+            }
+        }
+    }
+
+    private static Iterable<BlockPos> getBlocksForPoint(ColoredPoint loc) {
+        BlockPos minPos = new BlockPos(loc.vec3().x - 0.3D, loc.vec3().y - 1D, loc.vec3().z - 0.3D);
+        BlockPos maxPos = new BlockPos(loc.vec3().x + 0.3D, loc.vec3().y - 1D, loc.vec3().z + 0.3D);
+
+        return BlockPos.betweenClosed(minPos, maxPos);
+    }
+
+    public static int addNote(Component text) {
+        Entity root = McUtils.player().getRootVehicle();
+
+        LootrunUncompiled current = getActiveLootrun();
+
+        if (current == null) return 0;
+
+        current.notes().add(new ObjectObjectImmutablePair<>(root.position(), text));
+        return recompileLootrun(true);
+    }
+
+    @Nullable
+    public static LootrunUncompiled getActiveLootrun() {
+        LootrunUncompiled instance = null;
+        if (recording != null) instance = recording;
+        else if (uncompiled != null) instance = uncompiled;
+
+        return instance;
+    }
+
+    public static int recompileLootrun(boolean saveToFile) {
+        if (recording != null) {
+            recordingInformation.setDirty(true);
+        } else if (uncompiled != null) {
+            lootrun = compile(uncompiled);
+            if (saveToFile && uncompiled.file() != null) {
+                LootrunSaveResult lootrunSaveResult =
+                        saveLootrun(uncompiled.file().getName());
+                switch (lootrunSaveResult) {
+                    case SAVED -> {
+                        return 1;
+                    }
+                    case ERROR_SAVING, ERROR_ALREADY_EXISTS -> {
+                        return 0;
+                    }
+                }
+            }
+        }
+        return 1;
+    }
+
+    public static LootrunSaveResult saveLootrun(String name) {
```

Comment:
imo this could be a method of `LootrunUncompiled`, and on that note I feel as though `LootrunInstance` and `LootrunUncompiled` could be merged into one class, and that `RecordingInformation`'s `dirty` could also be situated there

Reply:
I would not merge the two, but I agree that methods can be moved.